src/server/transaction.cc

```diff
@@ -71,12 +71,58 @@ void RecordTxScheduleStats(const Transaction* tx) {
   }
 }
 
+void RecordTxScheduleFastStats(const Transaction* tx, bool was_ooo, bool was_inline) {
+  DCHECK_EQ(tx->GetUniqueShardCnt(), 1u);
+  auto* ss = ServerState::tlocal();
+  if (was_ooo) {
+    ss->stats.tx_type_cnt[was_inline ? ServerState::INLINE : ServerState::QUICK]++;
+  } else {
+    ss->stats.tx_type_cnt[ServerState::NORMAL]++;
+  }
+  ss->stats.tx_width_freq_arr[0]++;
+}
+
 }  // namespace
 
 IntentLock::Mode Transaction::LockMode() const {
   return cid_->IsReadOnly() ? IntentLock::SHARED : IntentLock::EXCLUSIVE;
 }
 
+uint32_t Transaction::PhasedBarrier::Start(uint32_t count) {
+  DCHECK_EQ(DEBUG_Get(), 0u);
+  count_.store(count, memory_order_release);
+  return phase_.load(memory_order_relaxed);
+}
+
+bool Transaction::PhasedBarrier::Active() const {
+  return count_.load(memory_order_acquire) > 0;
+}
+
+void Transaction::PhasedBarrier::Dec(Transaction* keep_alive) {
+  // Prevent transaction from being destroyed after count was decreased and Wait() unlocked,
+  // but before this thread finished notifying.
+  ::boost::intrusive_ptr guard(keep_alive);
+
+  uint32_t before = count_.fetch_sub(1);
+  CHECK_GE(before, 1u);
+  if (before == 1)
+    ec_.notify();
+}
+
+void Transaction::PhasedBarrier::Wait() {
+  ec_.await([this] { return count_.load(memory_order_acquire) == 0; });
```

Comment:
it's a bug and very subtle one. You can not do this:
```
Tthread1:  cond = true; ec->notify();
Thread2: ec->await(cond); .... delete ec;
```

There are several reasons, one of them thread1 can execute `cond=true`, then thread2 will execute await, continue deleting ec and then thread1 will have data race.

Even more subtle, notify itself wakes up thread2 but it still references ec one line after, so it can be and actually happens that thread 2 wakes up, deletes ec and then thread1 has a data race,

This is why Done wraps eventcount with reference count

Reply:
This is why I asked https://github.com/dragonflydb/dragonfly/pull/2455#discussion_r1461749346

It's the same code, but wrapped in a struct. I don't agree there is a bug here, because we still rely on use_count. 

I agree it's tricky. Once our guard from `Dec()` drops, the object can be gone, _while_ we're exiting the function. But we're been doing the same thing before with `DecreaseRunCnt()`

Reply:
I can refactor it into a static function so it's explicit that the object it not accessed anymore, but we can still be deleted from inside the transaction when we call `Dec()` instead of `DecreaseRunCnt()`

Reply:
https://isocpp.org/wiki/faq/freestore-mgmt#delete-this

Reply:
no need to refactor if it works well like this